src/rust/engine/ui/src/console_ui.rs

```diff
@@ -204,40 +117,324 @@ impl ConsoleUI {
+      // We acquire the lock before taking exclusive access, and don't release it until after
+      // initialization. That way, the exclusive callback can always assume that the destination
+      // is initialized (i.e., can `unwrap` it).
+      let mut stderr_dest_bar_guard = stderr_dest_bar.lock();
+      let (term_read, _, term_stderr_write) = {
+        let stderr_dest_bar = stderr_dest_bar.clone();
+        stdio::get_destination().exclusive_start(Box::new(move |msg: &str| {
+          // Acquire a handle to the destination bar in the UI. If we fail to upgrade, it's because
+          // the UI has shut down: we fail the callback to have the logging module directly log to
+          // stderr at that point.
+          let dest_bar = {
+            let stderr_dest_bar = stderr_dest_bar.lock();
+            // We can safely unwrap here because the Mutex is held until the bar is initialized.
+            stderr_dest_bar.as_ref().unwrap().upgrade().ok_or(())?
+          };
+          dest_bar.println(msg);
+          Ok(())
+        }))?
+      };
+
+      let stderr_use_color = term_stderr_write.use_color;
+      let term = console::Term::read_write_pair_with_style(
+        term_read,
+        term_stderr_write,
+        console::Style::new().force_styling(stderr_use_color),
+      );
+      // NB: We render more frequently than we receive new data in order to minimize aliasing where a
+      // render might barely miss a data refresh.
+      let draw_target = ProgressDrawTarget::term(term, ConsoleUI::render_rate_hz() * 2);
+      let multi_progress = MultiProgress::with_draw_target(draw_target);
+      let bars = (0..cmp::min(local_parallelism, terminal_height.into()))
+        .map(|_n| {
+          let style = ProgressStyle::default_bar().template("{spinner} {wide_msg}");
+          multi_progress.add(ProgressBar::new(terminal_width.into()).with_style(style))
+        })
+        .collect::<Vec<_>>();
+      *stderr_dest_bar_guard = Some(bars[0].downgrade());
+
+      // Spawn a task to drive the multi progress.
+      let multi_progress_task = executor
+        .spawn_blocking(move || multi_progress.join())
+        .boxed();
+
+      Ok(Instance::Indicatif(IndicatifInstance {
+        tasks_to_display: IndexSet::new(),
+        multi_progress_task,
+        bars,
+      }))
+    };
+  }
+
+  pub fn render(&mut self, heavy_hitters: &HashMap<SpanId, (String, SystemTime)>) {
+    let classify_tasks = |mut current_ids: HashSet<SpanId>,
+                          handler: &mut dyn FnMut(SpanId, TaskState)| {
+      for span_id in heavy_hitters.keys() {
+        let update = current_ids.remove(span_id);
+        handler(
+          *span_id,
+          if update {
+            TaskState::Update
+          } else {
+            TaskState::New
+          },
+        )
+      }
+      for span_id in current_ids {
+        handler(span_id, TaskState::Remove);
+      }
+    };
```

Comment:
It's way out of scope for this PR, but honestly `heavy_hitters()` could be refactored to be more like this function, where it simply calls into a function param with the relevant info.

Likely to be a performance bump from not constructing and moving the HashMap.

Reply:
Yea, there is absolutely optimization that can happen here. But I expect that it will be easier to do once we've reduced back to one implementation: `prodash`'s datastructure _might_ mean that we don't even need to keep the `running` graph... or can directly keep `prodash` nodes _in_ the running graph.

Reply:
Ah yes, that would truly be ideal. :raised_hands: